third_party/move/move-compiler-v2/src/pipeline/reference_safety_processor.rs

```diff
@@ -0,0 +1,1453 @@
+//! connected by an edge labeled with `f`. The borrow graph is a DAG (acyclic).
+//!
+//! Together with the borrow graph, a mapping from temporaries to graph nodes is maintained at
+//! each program point. These represent the currently active references into the borrowed data.
+//! All the _ancestors_ of those nodes are indirectly borrowed as well. Consider again `s` a
+//! struct, and assume ` let $t = &s.f` a field selection, then `$t` points to the node
+//! representing the reference to the field. However, the original `s` from which the field
+//! was selected, is also (indirectly) borrowed, as described by the borrow graph.
+//!
+//! The safety analysis essentially evaluates each instruction under the viewpoint of the current
+//! active borrow graph at the program point, to detect any conditions of non-safety. This
+//! includes specifically the following rules:
+//!
+//! (a) If immutable references to a location exist at a given program point, no mutable
+//!     references can co-exist.
+//! (b) Only one mutable reference can exist at time for the same location.
+//! (c) A location which is borrowed cannot be updated, and the value in it cannot be moved out.
+//! (d) References returned from a function call most be derived from parameters
+//!
+//! Hereby, selection of fields leads to independent 'sub-locations': the above rules are
+//! applying independently for such sub-locations. Thus one can have `&s.f` and `&mut s.g` at the
+//! same time.
+
+use crate::pipeline::livevar_analysis_processor::{LiveVarAnnotation, LiveVarInfoAtCodeOffset};
+use codespan_reporting::diagnostic::Severity;
+use im::ordmap::Entry;
+use itertools::Itertools;
+use move_binary_format::file_format::CodeOffset;
+use move_model::{
+    ast::TempIndex,
+    model::{FieldId, FunctionEnv, Loc, QualifiedInstId, StructId},
+    ty::Type,
+};
+use move_stackless_bytecode::{
+    dataflow_analysis::{DataflowAnalysis, TransferFunctions},
+    dataflow_domains::{AbstractDomain, JoinResult, MapDomain, SetDomain},
+    function_target::{FunctionData, FunctionTarget},
+    function_target_pipeline::{FunctionTargetProcessor, FunctionTargetsHolder},
+    stackless_bytecode::{AssignKind, AttrId, Bytecode, Operation},
+    stackless_control_flow_graph::StacklessControlFlowGraph,
+};
+use std::{
+    collections::{btree_map, BTreeMap, BTreeSet},
+    fmt::{Display, Formatter},
+};
+
+// ===============================================================================
+// Memory Safety Analysis
+
+// -------------------------------------------------------------------------------------------------
+// Program Analysis Domain
+
+/// Lifetime graph nodes represent information about memory locations, as well as nodes derived
+/// from that locations.
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct LifetimeNode {
+    /// Memory location associated with this node.
+    location: MemoryLocation,
+    /// Outgoing edges to descendants.
```

Comment:
Is this "descendants" or "children"?  Is it transitively closed?  If not, it should be "children".  And I suspect that instead of "ancestors" you mean "parents", unless that is also transitively closed.

Can you can add a comment to each field to tell us what is the interpretation of descendants and ancestors?  Ancestors has been previously mentioned, but this is the first use of the term "descendant".